### PR TITLE
Replace the HTML validator by a sanitizer based on lxml DOM

### PR DESCRIPTION
**Pull request by encukou**
As I'm working to decouple the content (forks) with its presentation (the website), I'm struggling with the fact that URL structure is baked into the HTML that comes out of the forks. I'd like to eventually add a way to adjust the links on naucse side.
That will require parsing the HTML. We already do that to validate it. Let's not do it twice.

This replaces the validator with a "sanitizer" that currently does the same thing, but will allows adjusting the HTML.

Another benefit should be that as we grow better validation, we'll be able to *remove* bad HTML from pre-existing forks, rather than stop rendering the fork or be forced to change it.

### CODE REVIEW

File: naucse/sanitize.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    elif element.tag is lxml.etree.Comment:
+        pass
+    else:
+        raise DisallowedElement(element)
+
+    # Allow only known attributes
+    for attr_name, value in list(element.items()):
+        if attr_name == 'style':
+            # Allow inline styles
+            # XXX: This is a potential security risk.
```
Comment (hroncok) [inline naucse/sanitize.py]: @encukou `sanitize_css` inline css maybe?

Comment (encukou) [reply naucse/sanitize.py]: Out of scope for this PR. The current validator allows CSS through (mainly for Notebook). This PR just has more prominent warnings around it.

File: test_naucse/test_validation.py [hunk 2]
```diff
@@ -1,100 +1,116 @@
...
 
-def test_allow_attributes():
-    allowed_elements = naucse.validation.AllowedElementsParser()
-
-    allowed_elements.reset_and_feed(
-        "<div class='test'><span style='color: red'><a href='/courses/'>Text</a></span></div>"
+def test_allowed_elements():
+    assert_changed(
+        "<div><strong><u><a>Test</a></u></div>",
+        "<div><strong><u><a>Test</a></u></strong></div>",
```
Comment (hroncok) [inline test_naucse/test_validation.py]: where is this change coming form? I might have just missed it, but I don't get it. si it a side effect of parsing HTML with lxml and dumping it back to string?

File: naucse/sanitize.py (hunk 1 above)
Comment (hroncok) [reply naucse/sanitize.py]: that was a not for the future

File: test_naucse/test_validation.py [hunk 3]
```diff
@@ -1,100 +1,116 @@
...
+
+def test_disallow_onhover():
+    with pytest.raises(naucse.sanitize.DisallowedAttribute):
+        sanitize_html("""
+            <div class="test" onhover="alert('XSS')">
+                <a href="/courses/">Text</a>
+            </div>
+        """)
+
+@pytest.mark.xfail
```
Comment (hroncok) [inline test_naucse/test_validation.py]: this is xfail for the XXX?

Comment (encukou) [reply test_naucse/test_validation.py]: Yes.

File: test_naucse/test_validation.py (hunk 2 above)
Comment (encukou) [reply test_naucse/test_validation.py]: Yes, this comes from the lxml round-trip.

Comment (hroncok) [reply test_naucse/test_validation.py]: consider adding a comment about it, as it is not so obvious.

**Comment by encukou** [comment]
Thank you for the review!

**Pull request merged**